### ISSUE TITLE
[Pilot] Apple ID or Password was entered incorrectly

### ISSUE THREAD

**Issue by squarefrog** [fastlane/fastlane#3876]
### New Issue Checklist

- [x] Updated fastlane to the latest version
- [x] I have read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)

### Issue Description

When attempting to deploy a new build of our app, fastlane is reporting an error during `pilot` submission.

```
[13:54:46]: [Transporter Error Output]: Your Apple ID or password was entered incorrectly. (-20101)
The login credentials for 'name@email.com' seem to be wrong
```

##### Complete output when running fastlane, including the stack trace and command used

I can't supply the complete stack trace due to personal information, but you should have it tagged: 4a831aa2e64d0f156038f759e59a68c9.

I have also tried adding my password into my environment:

```
ENV["FASTLANE_PASSWORD"] = "password"
```

I know the Apple ID and Password are correct as I've just used them to log into iTunes Connect, and my `Fastfile` hasn't changed in a while, and worked correctly last time. If I update the credentials when prompted, then try again the error is repeated.

### Configuration Files

Please copy the complete content of your `Fastfile` and any other configuration files you use below:

Fastfile:
```ruby
desc "Deploy a new version to the App Store"
lane :release do
  # get version number from file
  latest_version = get_latest_version 

  deliver(
    app_version: latest_version,
    skip_screenshots: true
  )

  beta(skip_bump: true)

  Actions.sh('git push --all --follow-tags')
end

desc "Submit a new Beta Build to Apple TestFlight"
desc "- this should be a weekly build"
lane :beta do |options|
  if !options[:skip_bump]
    ensure_git_branch(branch: 'develop')
    # bump version/build numbers
    bump 
  end

  fetch_certificates

  ipa = gym
  changelog = generate_changelog # load changelog text from file
  
  # This is where the incorrect password appears.
  pilot(
    ipa: ipa,
    changelog: changelog,
    skip_submission: true
  )
end
```

### Environment

fastlane version (fastlane -v): 1.67.0
Do you use bundler, rbenv or rvm? No

**Comment by squarefrog** [comment]
It's worthy of note that `pilot` logs in successfully, but later fails with incorrect password. The account used to log in is an admin account so shouldn't be a permissions issue.

```
[14:10:08]: ---------------------------------------------------
[14:10:13]: --- Step: Switch to ios generate_changelog lane ---
[14:10:13]: ---------------------------------------------------
[14:10:13]: Cruising over to lane 'ios generate_changelog' 🚖
[14:10:13]: Cruising back to lane 'ios beta' 🚘
[14:10:13]: -------------------
[14:10:13]: --- Step: pilot ---
[14:10:13]: -------------------
[14:10:13]: Login to iTunes Connect (name@email.com)
[14:10:15]: Login successful
[14:10:16]: Ready to upload new build to TestFlight (App: 963139521)...
[14:10:16]: Going to upload updated app to iTunes Connect
[14:10:16]: This might take a few minutes, please don't interrupt the script
[14:10:19]: [Transporter]: INFO: Transporter is skipping bundle update check: Already checked recently.
[14:10:19]: [Transporter]: INFO: Logging configured successfully.
[14:10:19]: [Transporter]: DEBUG: Attempting refresh of configuration data from https://contentdelivery.itunes.apple.com/transporter/Defaults.properties
[14:10:19]: [Transporter]: DEBUG: Configuration refresh successful.
[14:10:19]: [Transporter]: DEBUG: Saving configuration to local path: /Users/pi/.itmstransporter/Defaults.properties
[14:10:19]: [Transporter]: INFO: iTMSTransporter : iTunes Store Transporter [1.8.8]
[14:10:19]: [Transporter]: INFO: OS identifier: Mac OS X 10.11.4 (x86_64); jvm=24.80-b07; jre=1.7.0-internal-root_2015_05_12_09_52-b00
[14:10:19]: [Transporter]: INFO: Memory: [JVM] 889M free, 982M total, 982M max [System] (Physical) 520M free, 16384M total (Swap) 0 free, 0 total
[14:10:19]: [Transporter]: INFO: Upload mode selected.
[14:10:19]: [Transporter]: INFO: Examining the package at: /tmp/963139521.itmsp
[14:10:19]: [Transporter]: INFO: Ensuring that package has well formed metadata file...
[14:10:19]: [Transporter]: DEBUG: Getting case-sensitive metadata xml filename...
[14:10:19]: [Transporter]: INFO: Gathering the list of valid files from the package ...
[14:10:19]: [Transporter]: DEBUG:   515259d8a578db426fb41934eb40c50e.ipa will be verified by Apple's web service to determine if it is a valid file.
[14:10:19]: [Transporter]: DEBUG:   metadata.xml will be verified by Apple's web service to determine if it is a valid file.
[14:10:19]: [Transporter]: INFO: Finished gathering the list of valid files from the package.
[14:10:19]: [Transporter]: DEBUG:   Case-sensitive metadata filename is: metadata.xml
[14:10:19]: [Transporter]: INFO: Performing authentication of package 963139521.itmsp ...
[14:10:19]: [Transporter]: DBG-X: Memory: [JVM] 879M free, 982M total, 982M max [System] (Physical) 507M free, 16384M total (Swap) 0 free, 0 total
[14:10:19]: [Transporter]: DBG-X: Memory: [JVM] 879M free, 982M total, 982M max [System] (Physical) 507M free, 16384M total (Swap) 0 free, 0 total
[14:10:19]: [Transporter]: DBG-X: Using operation named: authenticateForSession
[14:10:19]: [Transporter]: DBG-X: Apple's web service operation input parameters:
[14:10:19]: [Transporter]: DBG-X:   parameter Application = iTMSTransporter
[14:10:19]: [Transporter]: DBG-X:   parameter BaseVersion = 1.8.8
[14:10:19]: [Transporter]: DBG-X:   parameter OSIdentifier = Mac OS X 10.11.4 (x86_64); jvm=24.80-b07; jre=1.7.0-internal-root_2015_05_12_09_52-b00
[14:10:19]: [Transporter]: DBG-X:   parameter Password = **hidden value**
[14:10:19]: [Transporter]: DBG-X:   parameter TransporterArguments = -m upload -u name@email.com -p **hidden value** -f /tmp/963139521.itmsp -t Signiant -k 100000
[14:10:19]: [Transporter]: DBG-X:   parameter Username = name@email.com
[14:10:19]: [Transporter]: DBG-X:   parameter Version = 1.8.8
[14:10:19]: [Transporter]: DBG-X:   parameter iTMSTransporterMode = upload
[14:10:19]: [Transporter]: INFO: id = 20160322141019-008
[14:10:20]: [Transporter]: DBG-X: Apple's web service operation return value:
[14:10:20]: [Transporter]: DBG-X:   parameter ErrorMessage = Your Apple ID or password was entered incorrectly. (-20101)
[14:10:20]: [Transporter]: DBG-X:   parameter Errors = [Your Apple ID or password was entered incorrectly. (-20101)]
[14:10:20]: [Transporter]: DBG-X:   parameter ErrorCode = -20101
[14:10:20]: [Transporter]: DBG-X:   parameter Success = false
[14:10:20]: [Transporter]: DBG-X: Deflated 450 bytes to 392
[14:10:20]: [Transporter]: DBG-X: Using operation named: validateMetadata
[14:10:20]: [Transporter]: DBG-X: Apple's web service operation input parameters:
[14:10:20]: [Transporter]: DBG-X:   parameter Application = iTMSTransporter
[14:10:20]: [Transporter]: DBG-X:   parameter BaseVersion = 1.8.8
[14:10:20]: [Transporter]: DBG-X:   parameter Files = [515259d8a578db426fb41934eb40c50e.ipa, metadata.xml]
[14:10:20]: [Transporter]: DBG-X:   parameter Metadata = <?xml version="1.0" encoding="UTF-8"?>
[14:10:20]: [Transporter]: DBG-X:   parameter MetadataChecksum = cf6174808199012d472a1a528c2cb14f
[14:10:20]: [Transporter]: DBG-X:   parameter MetadataCompressed = (suppressed)
[14:10:20]: [Transporter]: DBG-X:   parameter OSIdentifier = Mac OS X 10.11.4 (x86_64); jvm=24.80-b07; jre=1.7.0-internal-root_2015_05_12_09_52-b00
[14:10:20]: [Transporter]: DBG-X:   parameter PackageName = 963139521.itmsp
[14:10:20]: [Transporter]: DBG-X:   parameter PackageSize = 4705490
[14:10:20]: [Transporter]: DBG-X:   parameter Password = **hidden value**
[14:10:20]: [Transporter]: DBG-X:   parameter Throttle = 100000
[14:10:20]: [Transporter]: DBG-X:   parameter TransporterArguments = -m upload -u name@email.com -p **hidden value** -f /tmp/963139521.itmsp -t Signiant -k 100000
[14:10:20]: [Transporter]: DBG-X:   parameter Username = name@email.com
[14:10:20]: [Transporter]: DBG-X:   parameter Version = 1.8.8
[14:10:20]: [Transporter]: DBG-X:   parameter iTMSTransporterMode = upload
[14:10:20]: [Transporter]: INFO: id = 20160322141020-923
[14:10:20]: [Transporter]: DBG-X: Apple's web service operation return value:
[14:10:20]: [Transporter]: DBG-X:   parameter ErrorMessage = Your Apple ID or password was entered incorrectly. (-20101)
[14:10:20]: [Transporter]: DBG-X:   parameter Errors = [Your Apple ID or password was entered incorrectly. (-20101)]
[14:10:20]: [Transporter]: DBG-X:   parameter ErrorCode = -20101
[14:10:20]: [Transporter]: DBG-X:   parameter Success = false
[14:10:20]: [Transporter Error Output]: Your Apple ID or password was entered incorrectly. (-20101)
The login credentials for 'name@email.com' seem to be wrong
Do you want to re-enter your password? (y/n)
```

**Note:** email address changed to name@email.com.

**Comment by TKBurner** [comment]
Hi @squarefrog . Can you try manually setting the 'username' flag for Pilot. 

pilot(
    ipa: ipa,
    changelog: changelog,
    skip_submission: true
    username: your@email.com
  )

Let me know if that clears things up :+1:

**Comment by squarefrog** [comment]
@TKBurner I would, but I've just changed my password to something without any special characters as a test, and it worked without exception.

Another peculiarity: I still had the old password environment variable in my `Fastfile`. When prompted to add my new password to `credentials_manager` it attempted to use my environment variable, which caused it to prompt me to add my password again in a loop. Removing the variable stopped this behaviour.

So it definitely seemed the presence of the ` { / ` characters in the old password caused `pilot` to fail.

**Comment by TKBurner** [comment]
Thanks for letting me know @samgreen. I'm going to loop in @mpirri to take a look.

**Comment by squarefrog** [comment]
No problem. Ping me if there's anything else you need.

**Comment by BrettThePark** [comment]
I also have this issue using a ! in the iTunes password.

**Comment by BrettThePark** [comment]
@TKBurner Setting the username in pilot does not seem to fix the issue.

**Comment by dusek** [comment]
Maybe solved by PR #3832 .

**Comment by TKBurner** [comment]
@BrettThePark After #3832 gets merged can you update and let me know if you issue is resolved? 

Thanks!

**Comment by mfurtak** [comment]
Hi folks - we've shipped [`fastlane_core` 0.41.1](https://github.com/fastlane/fastlane/releases/tag/fastlane_core%2F0.41.1) :rocket:  This contains an experimental fix for the iTMSTransporter password woes.

If you're interested in trying this fix and helping us validate it before we turn it on by default, please do the following:

1. Run `gem update fastlane_core` and be sure that you get **0.41.1**
1. Get your favorite strong password set up for your Apple ID and ensure that any stored Keychain passwords match
1. Run `fastlane`, `pilot`, or `deliver` as normal, but put `FASTLANE_EXPERIMENTAL_TRANSPORTER_AVOID_SHELL_SCRIPT=1` before the whole command
    * e.g. `FASTLANE_EXPERIMENTAL_TRANSPORTER_AVOID_SHELL_SCRIPT=1 fastlane beta`
    * Really long variable names just feel more arcane and powerful, don't you think? :wink:
1. Let us know what you see! We're hoping you're back in business.

There's some more information over at the [PR](https://github.com/fastlane/fastlane/pull/4003). Thanks! :raised_hands:

**Comment by davideme** [comment]
I had the same issue and with a password containing a (. 
And using `FASTLANE_EXPERIMENTAL_TRANSPORTER_AVOID_SHELL_SCRIPT=1 fastlane beta`

it's working!

**Comment by fitch** [comment]
I seem to be getting the same error with `fastlane 1.103.0`. I have a `"` in my password.

**Comment by mfurtak** [comment]
Hi @fitch - If you're still having trouble, do you mind opening a new ticket and providing your setup info and output from your run? Thanks 👍
